sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -205,6 +226,117 @@ case class BroadcastNestedLoopJoinExec(
+            logWarning(s"failed to find notInSubquerySingleColumnOptimizeStreamedKeyIndex," +
+              s" fallback to leftExistenceJoin.")
+            false
+          }
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  private def notInSubquerySingleColumnOptimizeBuildParams(
+      buildRows: Array[InternalRow]): NotInSubquerySingleColumnOptimizeParams = {
+    val buildRowsHashSet = mutable.HashSet[AnyRef]()
+    val dataType = right.output.head.dataType
+    var isNullExist = false
+    buildRows.foreach(row => {
+      if (row.isNullAt(0)) {
+        isNullExist = true
+      }
+      // put null as value, because we only leverage HashMap keys
+      buildRowsHashSet.add(row.get(0, dataType))
+    })
+    NotInSubquerySingleColumnOptimizeParams(buildRowsHashSet, isNullExist, buildRows.isEmpty)
+  }
+
+  private def notInSubquerySingleColumnOptimizeSetStreamedKey(
+      leftAttr: AttributeReference,
+      rightAttr: AttributeReference): Unit = {
+    val leftNotInKeyFound = left.output.find(_.semanticEquals(leftAttr))
+    if (leftNotInKeyFound.isDefined) {
+      notInSubquerySingleColumnOptimizeStreamedKey = leftAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(leftAttr.exprId)
+    } else {
+      notInSubquerySingleColumnOptimizeStreamedKey = rightAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(rightAttr.exprId)
+    }
+  }
+
+  /**
+   * Optimized version implementation for LeftAnti
+   * which converted from single column NotInSubquery
+   * @param relation
+   * @return
+   */
+  private def notInSubquerySingleColumnOptimize(
+      relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    assert(buildSide == BuildRight)
+    val buildRows = relation.value
+    if (buildRows.length > SQLConf.get.notInSubquerySingleColumnOptimizeRowCountThreshold) {
+      logWarning(s"BuildSide row count ${buildRows.length} exceeded " +
+        s"notInSubquerySingleColumnOptimizeRowCountThreshold ${SQLConf.get
+          .notInSubquerySingleColumnOptimizeRowCountThreshold}, fallback to leftExistenceJoin.")
+      leftExistenceJoin(relation, false)
+    } else {
+      val params = notInSubquerySingleColumnOptimizeBuildParams(buildRows)
+      streamed.execute().mapPartitionsInternal { streamedIter =>
+        streamedIter.filter(row => {
+          // See. not-in-unit-tests-single-column.sql for detail filter rules
+          if (params.isBuildRowsEmpty) {
+            true
+          } else {
+            val streamedRowIsNull = row.isNullAt(notInSubquerySingleColumnOptimizeStreamedKeyIndex)
+            val streamedRowNotInKey = row.get(
+              notInSubquerySingleColumnOptimizeStreamedKeyIndex,
+              notInSubquerySingleColumnOptimizeStreamedKey.dataType
+            )
+            val notInKeyEqual = params.buildSideHashSet.contains(streamedRowNotInKey)
```

Comment:
Could we reuse `[Unsafe|Long]HashedRelation` here?

Reply:
done changing into HashedRelation, nice advise.